src/modules/Elsa.Workflows.Core/Features/WorkflowsFeature.cs

```diff
@@ -128,9 +129,11 @@ private void AddElsaCore(IServiceCollection services)
             .AddScoped(typeof(Func<IWorkflowBuilder>), sp => () => sp.GetRequiredService<WorkflowBuilder>())
             .AddScoped<IWorkflowBuilderFactory, WorkflowBuilderFactory>()
             .AddScoped<IVariablePersistenceManager, VariablePersistenceManager>()
-            .AddScoped<IIncidentStrategyResolver, DefaultIncidentStrategyResolver>()
-
-            // Incident Strategies.
+            .AddScoped<IIncidentStrategyResolver, DefaultIncidentStrategyResolver>()
+            .AddSingleton<ILoggerStateGenerator<WorkflowExecutionContext>, WorkflowLoggerStateGenerator>()

```

Comment:
@sfmskywalker Do you ever use TryAddSingleton? It seems like it could be useful here since otherwise the built-in implementation would override a custom implementation if already declared earlier. Or is it generally expected to add the override implementations after the .AddElsa()?

Reply:
In Elsa, the typical pattern is to expose a delegate property from the feature class, which allows other modules or the application itself to register custom service implementations.

For example, consider the following structure in `FeatureA`:

```csharp
public class FeatureA : FeatureBase
{
    public Func<IServiceProvider, IMyService> MyServiceFactory { get; set; } = sp => new DefaultMyService();

    // This method is invoked by the feature framework after all features have had a chance to configure themselves and other features.
    public override void Apply()
    {
        Services.AddSingleton(MyServiceFactory);
    }
}
```

In the specific scenario you mentioned, this pattern isn't being followed, but it could certainly be introduced.

Alternatively, we could argue that it's likely an edge case for someone to want to provide their own implementation of `ILoggerStateGenerator<T>`. In this case, instead of using `AddSingleton`, we could switch to `TryAddSingleton`. However, this would require careful consideration of the order in which features are registered. The developer would need to ensure that their `TryAddSingleton` call occurs before the one from `WorkflowsFeature`. If that's the case, the developer could just as easily invoke the feature early and use `AddSingleton`. In such a scenario, multiple service descriptors might exist for the same service, but the last one to be registered would take precedence.

My take: the most flexible way is to use the delegate pattern that the Features framework supports.